doc/manual/interfaces.rst

```diff
@@ -0,0 +1,223 @@
+
+Any object that has these three methods appropriately defined can be used in a ``for`` loop since the syntax::
+
+    for i in iter   # or  "for i = iter"
+      # body
+    end
+
+is translated into::
+
+    state = start(iter)
+    while !done(iter, state)
+      (i, state) = next(iter, state)
+      # body
+    end
+
+A simple example is an iterable collection of square numbers with a defined length::
+
+    immutable Squares
+        count::Int
+    end
+    Base.start(::Squares) = 1
+    Base.next(S::Squares, state) = (state*state, state+1)
+    Base.done(S::Squares, s) = s > S.count
+
+With only those definitions, the ``Squares`` type is already pretty powerful. We can iterate over all the elements::
+
+    julia> for i in Squares(10)
+               print(i, ", ")
+           end
+    1, 4, 9, 16, 25, 36, 49, 64, 81, 100,
+
+We can compute the sum of all squares up to a certain number::
+
+    julia> sum(Squares(1803))
+    1955361914
+
+Or even the mean and standard deviation::
+
+    julia> mean(Squares(100)), std(Squares(100))
+    (3383.5,3024.355854282583)
+
+There are a few more methods we can extend to give Julia more information about this iterable collection.  We know that the elements in a ``Squares`` collection will always be ``Int``. By extending the :func:`eltype` method, we can give that information to Julia and help it make more specialized code in the more complicated methods. We also know the number of elements in our collection, so we can extend :func:`length`, too::
+
+    Base.eltype(::Type{Squares}) = Int # Note that this is defined for the type
+    Base.length(S::Squares) = S.count
+
+Now, when we ask Julia to :func:`collect` all the elements into an array it can preallocate a ``Vector{Int}`` of the right size instead of blindly ``push!``\ ing each element into a ``Vector{Any}``::
+
+    julia> collect(Squares(100))' # transposed to save space
+    1x100 Array{Int64,2}:
+     1  4  9  16  25  36  49  64  81  100  …  9025  9216  9409  9604  9801  10000
+
+While we can rely upon generic implementations, we can also extend specific methods where we know there is a simpler algorithm.  For example, there's a formula to compute the sum of squares, so we can override the generic iterative version with a more performant solution::
+
+    julia> sum(S::Squares) = (n = S.count; return n*(n+1)*(2n+1)÷6)
```

Comment:
Hah, I was reading closely, I've been waiting for this document for a long time. Now we just need one for the show/print zoo...